### ISSUE TITLE
proposal: [apollo-datasource-rest] - bypass redis cache while redis host is not reachable

### ISSUE THREAD

**Issue by ecerroni** [apollographql/apollo-server#3469]
If Redis host is down, i.e. not reachable, the query response gets this error:
```
 "message": "Reached the max retries per request limit (which is 3). Refer to \"maxRetriesPerRequest\" option for details."
```

![redis-down](https://user-images.githubusercontent.com/11174327/68114521-84817f80-fef6-11e9-8c9b-c1b0f2269af0.png)

However, I think that showing this kind of error should be avoided and the default response's flow should be granted.

In `https://github.com/apollographql/apollo-server/blob/master/packages/apollo-datasource-rest/src/HTTPCache.ts` line 38 will assign the ioredis' error to `entry`.

```
...
const entry = await this.keyValueCache.get(cacheKey);
    if (!entry) {
      const response = await this.httpFetch(request);
      ...
```

I guess entry should be null in that case so that `const response = await this.httpFetch(request);` can be executed if redis' host is down.

**Comment by Yunoo** [comment]
I also think this is a better and more robust approach.
It is fine to get redis exceptions in case the host is down. However, the cache should be bypassed in that case.

**Comment by geraldstrobl** [comment]
@ecerroni did you open a PR or a workaround for this issue?

**Comment by ecerroni** [comment]
@geraldstrobl Nope, no PR.

This issue is been waiting for a response from the team. Still nothing.

I guess it is not a high priority.

**Comment by zordius** [comment]
Hope this to be fixed, it will be better for production usage.

**Comment by zordius** [comment]
This happened in memcached cache, too.

**Comment by gtothesquare** [comment]
Found a workround;
I end up implementing my own ApolloRedisCache using `apollo-server-caching` spec

```
const Redis = require('ioredis');

class MyApolloCache extends Redis {
  // set(key: string, value: V, options?: KeyValueCacheSetOptions): Promise<void>;
  async set(key, value, options = {}) {
    // check that redis is ready otherwise resolve with undefined to bypass
    if (this.status !== 'ready') {
      return undefined;
    }

    if (options.ttl) {
      return super.set(key, value, 'ex', options.ttl);
    }
    return super.set(key, value);
  }
  // get(key: string): Promise<V | undefined>;
  async get(key) {
    // check that redis is ready otherwise resolve with undefined to bypass
    if (this.status !== 'ready') {
      return undefined;
    }
    const value = await super.get(key);
    // ioredis returns null when not found however,
    // the spec `apollo-server-caching` states that we should return undefined
    if (value === null) {
      return undefined;
    }
    return value;
  }
}

module.exports = { MyApolloCache };

```
So install `apollo-server-caching` (mostly for the interfaces) and `ioredis` and create a 
enhanced version of `ioredis` (which also you will need to install as a separate package)

**Comment by Yunoo** [comment]
You could also extend the `RedisCache` class from the `apollo-server-cache-redis` package and override some of its functions to bypass cache when redis is unavailable. 

`import { RedisCache } from 'apollo-server-cache-redis'

export default class SafeRedisCache extends RedisCache {
  isReady() {
    if (!this.client || !this.client.status) return false
    return this.client.status.toLowerCase() === 'ready'
  }

  async set(key, value, options) {
    if (!this.isReady()) return undefined
    await super.set(key, value, options)
  }

  async get(key) {
    if (!this.isReady()) return undefined
    return super.get(key)
  }

  async delete(key) {
    if (!this.isReady()) return undefined
    return super.delete(key)
  }
}`

Then you can just `import RedisCache from './filename'` instead and use it in your ApolloServer.
The only catch is that the cache doesn't fallback to the LRU in-memory cache which is the default implementation of `apollo-datasource-rest`. Thus, no requests will be cached at all.

**Comment by timsuchanek** [comment]
In case you want to save yourself all of this overhead, here a shameless plug:
I built https://graphcdn.io, which takes care of the caching for you.
I promise you, that you won't hit these limitations there.

**Comment by trevor-scheer** [comment]
In case y'all are still following along (👋 2019), I've just published a couple caching utilities you might find handy for this reason (and possibly others!). I'd be interested to hear if my (simple) implementation is sufficient for your use case and open to feedback if it can be improved.

A couple things:
We've added a cache wrapper to allow easy integration the `keyv` package which supports a number of caching clients out of the box. (i.e. redis, memcached)
https://github.com/apollographql/apollo-utils/tree/main/packages/keyvAdapter#keyvadapter-class

I've also added an `ErrorsAreMissesCache` wrapper for Apollo-style `KeyValueCache` caches which are compatible with today's Apollo Server versions (and will be compatible with AS4 as well). More details and an example here:
https://github.com/apollographql/apollo-utils/tree/main/packages/keyValueCache#errorsaremissescache

The folks working on Apollo Server are moving away from offering various cache implementations in AS4, which is why we've made it easy to integrate with `keyv` (whose focus is exactly that!). If anyone ventures to try that out, I'd be very interested in hearing your experience using it.
